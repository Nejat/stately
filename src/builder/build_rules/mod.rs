pub use initial::*;
pub use state::*;
pub use transitions::*;
pub use triggers::*;

mod initial;
mod transitions;
mod triggers;
mod state;
