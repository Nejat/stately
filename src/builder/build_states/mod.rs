pub use builder::BuilderState;
pub use initial::InitialState;
pub use transition::{TransitionsState, TransitionState};
pub use trigger::{EndTriggersState, EndTriggerState, TriggersState, TriggerState};

mod builder;
mod initial;
mod transition;
mod trigger;
