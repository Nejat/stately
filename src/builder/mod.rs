pub use build_states::{
    BuilderState, EndTriggersState, EndTriggerState, InitialState,
    TransitionsState, TransitionState, TriggersState, TriggerState,
};
pub use builder::StateMachineBuilder;
pub use error::BuilderError;
pub use result::Result;

mod build_states;
#[allow(clippy::module_inception)]
mod builder;
mod error;
mod result;
