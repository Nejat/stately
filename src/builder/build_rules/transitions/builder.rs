use std::hash::Hash;

use crate::build_rules::StateBuilder;
use crate::builder::build_rules::MultiTransitionBuilder;
use crate::builder::result::Result;
use crate::StateMachineBuilder;

pub trait TransitionBuilder<TState, TEvent>
    where Self: Sized
{
    type MultiBuilder: MultiTransitionBuilder<TState, TEvent>;
    type StateBuilder: StateBuilder<TState, TEvent>;

    fn transition_on(
        self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::MultiBuilder, TState, TEvent>;

    fn only_transition_on(
        self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::StateBuilder, TState, TEvent>;
}

impl<TState, TEvent> TransitionBuilder<TState, TEvent> for StateMachineBuilder<TState, TEvent>
    where TState: Copy + Eq + Hash,
          TEvent: Eq + Hash,
{
    type MultiBuilder = StateMachineBuilder<TState, TEvent>;
    type StateBuilder = StateMachineBuilder<TState, TEvent>;

    #[inline]
    fn transition_on(
        mut self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::MultiBuilder, TState, TEvent> {
        self.add_transition_impl(self.current, event, next_state).map(|_| self)
    }

    #[inline]
    fn only_transition_on(
        mut self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::StateBuilder, TState, TEvent>
    {
        self.add_transition_impl(
            self.current,
            event,
            next_state,
        ).map(|_| self as StateMachineBuilder<TState, TEvent>)
    }
}
