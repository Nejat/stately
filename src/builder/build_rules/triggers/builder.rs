use std::hash::Hash;

use crate::builder::build_rules::{MultiTriggerBuilder, TransitionBuilder, TriggerEndBuilder};
use crate::StateMachineBuilder;

pub trait TriggerBuilder<TState, TEvent> {
    type EndBuilder: TriggerEndBuilder<TState, TEvent>;
    type MultiBuilder: MultiTriggerBuilder<TState, TEvent>;
    type TransitionBuilder: TransitionBuilder<TState, TEvent>;

    #[must_use]
    fn also_end_state(self) -> Self::EndBuilder;

    #[must_use]
    fn trigger(self, trigger: impl Fn(TEvent, TState, TState) + 'static) -> Self::MultiBuilder;

    #[must_use]
    fn no_triggers(self) -> Self::TransitionBuilder;

    #[must_use]
    fn only_trigger(self, trigger: impl Fn(TEvent, TState, TState) + 'static) -> Self::TransitionBuilder;
}

impl<TState, TEvent> TriggerBuilder<TState, TEvent> for StateMachineBuilder<TState, TEvent>
    where TState: Copy + Eq + Hash,
          TEvent: Eq + Hash,
{
    type EndBuilder = StateMachineBuilder<TState, TEvent>;
    type MultiBuilder = StateMachineBuilder<TState, TEvent>;
    type TransitionBuilder = StateMachineBuilder<TState, TEvent>;

    fn also_end_state(mut self) -> Self::EndBuilder {
        self.also_end_state_impl(self.current);

        self
    }

    #[inline]
    fn trigger(mut self, trigger: impl Fn(TEvent, TState, TState) + 'static) -> Self::MultiBuilder {
        self.trigger_on_impl(self.current, trigger);

        self
    }

    #[inline]
    fn no_triggers(self) -> Self::TransitionBuilder {
        self
    }

    #[inline]
    fn only_trigger(mut self, trigger: impl Fn(TEvent, TState, TState) + 'static) -> Self::TransitionBuilder {
        self.trigger_on_impl(self.current, trigger);

        self
    }
}
