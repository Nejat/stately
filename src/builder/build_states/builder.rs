use std::collections::HashSet;
use std::hash::Hash;
use std::rc::Rc;

use crate::builder::{EndTriggerState, TriggerState};
use crate::builder::builder::{NodeType, StateMachineBuilder};
use crate::builder::BuilderError::ValidationError;
use crate::builder::Result;
use crate::StateMachineDefinition;

pub trait BuilderState<TState, TEvent> {
    type EndState: EndTriggerState<TState, TEvent>;

    type TriggerState: TriggerState<TState, TEvent>;

    fn add_end_state(self, state: TState) -> Result<Self::EndState, TState, TEvent>;

    fn add_start_state(
        self,
        event: TEvent,
        state: TState,
    ) -> Result<Self::TriggerState, TState, TEvent>;

    fn add_start_end_state(
        self,
        event: TEvent,
        state: TState,
    ) -> Result<Self::EndState, TState, TEvent>;

    fn add_state(self, state: TState) -> Result<Self::TriggerState, TState, TEvent>;

    fn build(self) -> Result<StateMachineDefinition<TState, TEvent>, TState, TEvent>;
}

impl<TState, TEvent> BuilderState<TState, TEvent> for StateMachineBuilder<TState, TEvent>
    where TState: Copy + Eq + Hash,
          TEvent: Eq + Hash,
{
    type EndState = Self;
    type TriggerState = Self;

    #[inline]
    fn add_end_state(mut self, state: TState) -> Result<Self::TriggerState, TState, TEvent> {
        self.add_state_impl(state, NodeType::END).map(|_| self)
    }

    #[inline]
    fn add_start_state(
        mut self,
        event: TEvent,
        state: TState,
    ) -> Result<Self::TriggerState, TState, TEvent> {
        self.add_start_state_impl(event, state).map(|_| self)
    }

    fn add_start_end_state(mut self, event: TEvent, state: TState) -> Result<Self::EndState, TState, TEvent> {
        self.add_start_end_state_impl(event, state).map(|_| self)
    }

    #[inline]
    fn add_state(mut self, state: TState) -> Result<Self::TriggerState, TState, TEvent> {
        self.add_state_impl(state, NodeType::STATE).map(|_| self)
    }

    fn build(self) -> Result<StateMachineDefinition<TState, TEvent>, TState, TEvent> {
        let undefined_states = self.transitions.iter()
            .flat_map(|(_, states)| states.values().collect::<Vec<_>>())
            .filter(|state| !self.states.contains(state))
            .copied()
            .collect::<HashSet<_>>().into_iter()
            .collect::<Vec<_>>();

        let unreachable = self.states.iter()
            .filter(|state| self.transitions.iter()
                .all(|(from, transitions)|
                    from == *state || transitions.values().all(|next| { next != *state })
                ))
            .copied()
            .collect::<Vec<_>>();

        if undefined_states.is_empty() && unreachable.is_empty() {
            Ok(StateMachineDefinition {
                end_states: Rc::new(self.end_states),
                initial_state: self.initial_state,
                transitions: Rc::new(self.transitions),
                triggers: Rc::new(self.triggers),
            })
        } else {
            Err(ValidationError {
                undefined_states,
                unreachable,
            })
        }
    }
}
