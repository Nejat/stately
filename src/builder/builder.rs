use std::borrow::BorrowMut;
use std::collections::{HashMap, HashSet};
use std::hash::Hash;

use bitflags::bitflags;

use crate::{builder, Trigger};
use crate::builder::BuilderError::{
    RedefinedInitialState, StateAlreadyDefined, TransitionAlreadyDefined,
};
use crate::builder::InitialState;

bitflags! {
    #[derive(Copy, Clone)]
    pub struct NodeType: u8 {
        const END   = 0b10;
        const START = 0b01;
        const STATE = 0b00;
    }
}

/// Fluent state machine builder, which is used to build a validated
/// [`StateMachineDefinition`]
///
/// Validation occurs in three separate phases;
///
/// 1. Transitioning between state builder states to eliminate a class of validations
/// enforced at compiletime
/// 2. Defining different components of the state machine definition
/// 3. Building the state machine definition; where unreachable and undefined
/// states are detected
///
/// ### Example
///
///  <img src="https://raw.githubusercontent.com/Nejat/stately/master/design/diagrams/cyclical.svg" alt="cyclical states" width="400"
///       style="background: transparent; position: absolute; Left: 450px; Margin-Top: 400px; z-index: 10000"/>
///
/// ```rust
/// use std::fmt::{Display, Formatter};
///
/// use stately::builder::Result;
/// use stately::prelude::*;
///
/// use Event::*;
/// use State::*;
///
/// type BuilderResult = Result<StateMachineDefinition<State, Event>, State, Event>;
///
/// fn cyclical_fsm() -> BuilderResult {
///     StateMachineBuilder::new()
///         .add_start_state(Start, A)?
///             .no_triggers()
///             .transition_on(Done, E)?
///             .final_transition_on(Next, B)?
///         .add_end_state(E)?
///             .no_triggers()
///         .add_state(B)?
///             .no_triggers()
///             .transition_on(Next, C)?
///             .transition_on(Loop, B1)?
///             .final_transition_on(Done, G)?
///         .add_state(B1)?
///             .no_triggers()
///             .only_transition_on(Next, D)?
///         .add_state(C)?
///             .no_triggers()
///             .only_transition_on(Next, D)?
///         .add_state(D)?
///             .no_triggers()
///             .transition_on(Next, F)?
///             .transition_on(Loop, B)?
///             .final_transition_on(Done, G)?
///         .add_end_state(F)?
///             .no_triggers()
///         .add_end_state(G)?
///             .no_triggers()
///         .add_start_end_state(Skip, H)?
///             .no_triggers()
///         .build()
/// }
///
/// #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
/// pub enum Event {
///     Done,
///     Loop,
///     Next,
///     Skip,
///     Start,
/// }
///
/// #[derive(Copy, Clone, Default, Debug, Eq, PartialEq, Hash)]
/// enum State {
///     #[default]
///     Initial,
///     A,
///     B,
///     B1,
///     C,
///     D,
///     E,
///     F,
///     G,
///     H,
/// }
/// ```
///
/// [`StateMachineDefinition`]: crate::state_machine::StateMachineDefinition
pub struct StateMachineBuilder<TState, TEvent> {
    pub(crate) current: TState,
    pub(crate) initial_state: TState,
    pub(crate) end_states: HashSet<TState>,
    pub(crate) start_states: HashSet<TState>,
    pub(crate) states: HashSet<TState>,
    pub(crate) transitions: HashMap<TState, HashMap<TEvent, TState>>,
    pub(crate) triggers: HashMap<TState, Vec<Trigger<TState, TEvent>>>,
}

impl<TState, TEvent> StateMachineBuilder<TState, TEvent>
    where TState: Copy + Default + Eq + Hash,
          TEvent: Copy + Eq + Hash
{
    /// Create a new instance of a `StateMachineBuilder`
    ///
    /// # Results
    ///
    /// Returns an [`InitialState`] to start the build process
    ///
    #[allow(clippy::new_ret_no_self)]
    #[must_use]
    pub fn new() -> impl InitialState<TState, TEvent> {
        let initial_state = TState::default();

        Self {
            current: initial_state,
            initial_state,
            end_states: HashSet::default(),
            start_states: HashSet::default(),
            states: HashSet::default(),
            transitions: HashMap::default(),
            triggers: HashMap::default(),
        }
    }
}

impl<TState, TEvent> StateMachineBuilder<TState, TEvent>
    where TState: Copy + Eq + Hash,
          TEvent: Eq + Hash
{
    /// Defines a new state for the state machine
    ///
    /// _used by specific trait implementations of the state machine builder_
    ///
    /// # Arguments
    ///
    /// * _`state`_ - state to add
    /// * _`NodeType`_ - node type of the state being added; _i.e start, end, start|end_
    ///
    /// # Results
    ///
    /// Returns `()` if there aren't any validation errors
    ///
    /// # Errors
    ///
    /// Returns a [`BuildError`] if there are any validation errors
    ///
    /// [`BuildError`]: builder::BuilderError
    pub(crate) fn add_state_impl(
        &mut self,
        state: TState,
        node_type: NodeType,
    ) -> builder::Result<(), TState, TEvent> {
        // check if redefining `initial` state
        if state == self.initial_state {
            return Err(RedefinedInitialState);
        }

        // check if redefining a state that was already defined
        if self.states.contains(&state) {
            return Err(StateAlreadyDefined { state });
        }

        // add state to tracker
        self.states.borrow_mut().insert(state);

        // if the state is an end state, define it as such
        if node_type.contains(NodeType::END) {
            self.end_states.borrow_mut().insert(state);
        }

        // if the state is a start state, define it as such
        if node_type.contains(NodeType::START) {
            self.start_states.borrow_mut().insert(state);
        }

        // record the new state as the current state
        self.current = state;

        Ok(())
    }

    /// Defines a new start/end state for the state machine
    ///
    /// _used by specific trait implementations of the state machine builder_
    ///
    /// # Arguments
    ///
    /// * `event` - the event that starts the state machine
    /// * `state` - the start/end state that the state machine transitions to
    ///
    /// # Results
    ///
    /// Returns `()` if there aren't any validation errors
    ///
    /// # Errors
    ///
    /// Returns a [`BuildError`] if there are any validation errors
    ///
    /// [`BuildError`]: builder::BuilderError
    #[inline]
    pub(crate) fn add_start_end_state_impl(
        &mut self,
        event: TEvent,
        state: TState,
    ) -> builder::Result<(), TState, TEvent> {
        self.add_state_impl(state, NodeType::START | NodeType::END)?;
        self.add_transition_impl(self.initial_state, event, state)
    }

    /// Defines a new start state for the state machine
    ///
    /// _used by specific trait implementations of the state machine builder_
    ///
    ///
    /// # Arguments
    ///
    /// * `event` - the event that starts the state machine
    /// * `state` - the start state that the state machine transitions to
    ///
    /// # Results
    ///
    /// Returns `()` if there aren't any validation errors
    ///
    /// # Errors
    ///
    /// Returns a [`BuildError`] if there are any validation errors
    ///
    /// [`BuildError`]: builder::BuilderError
    #[inline]
    pub(crate) fn add_start_state_impl(
        &mut self,
        event: TEvent,
        state: TState,
    ) -> builder::Result<(), TState, TEvent> {
        self.add_state_impl(state, NodeType::START)?;
        self.add_transition_impl(self.initial_state, event, state)
    }

    /// Defines a new transition for the current state being defined
    /// for the state machine
    ///
    /// _used by specific trait implementations of the state machine builder_
    ///
    /// # Arguments
    ///
    /// * _`state`_ - the event that transitions the state machine
    /// * _`event`_ - the state that the state machine transitions from
    /// * _`next`_ - the state that the state machine transitions to
    ///
    /// # Results
    ///
    /// Returns `()` if there aren't any validation errors
    ///
    /// # Errors
    ///
    /// Returns a [`BuildError`] if there are any validation errors
    ///
    /// [`BuildError`]: builder::BuilderError
    pub(crate) fn add_transition_impl(
        &mut self,
        state: TState,
        event: TEvent,
        next: TState,
    ) -> builder::Result<(), TState, TEvent> {
        let entry = self.transitions.entry(state)
            .or_insert_with(HashMap::new);

        if entry.contains_key(&event) {
            let existing = entry[&event];

            return Err(TransitionAlreadyDefined { event, existing });
        }

        entry.borrow_mut().insert(event, next);

        Ok(())
    }

    /// Defines a new trigger for the current state being defined
    /// for the state machine
    ///
    /// _used by specific trait implementations of the state machine builder_
    ///
    /// # Arguments
    ///
    /// * _`state`_ - the state that the state machine transitions to
    /// * _`trigger`_ - the callback to invoke on transition
    ///
    pub(crate) fn trigger_on_impl(
        &mut self,
        state: TState,
        trigger: impl Fn(TEvent, TState, TState) + 'static,
    ) {
        self.triggers.entry(state)
            .or_insert_with(Vec::new)
            .push(Box::new(trigger));
    }
}
