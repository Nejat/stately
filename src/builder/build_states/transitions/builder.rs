use std::hash::Hash;

use crate::builder::build_states::TransitionsState;
use crate::builder::BuilderState;
use crate::builder::Result;
use crate::StateMachineBuilder;

pub trait TransitionState<TState, TEvent>
    where Self: Sized
{
    type MultiState: TransitionsState<TState, TEvent>;
    type BuildState: BuilderState<TState, TEvent>;

    fn transition_on(
        self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::MultiState, TState, TEvent>;

    fn only_transition_on(
        self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::BuildState, TState, TEvent>;
}

impl<TState, TEvent> TransitionState<TState, TEvent> for StateMachineBuilder<TState, TEvent>
    where TState: Copy + Eq + Hash,
          TEvent: Eq + Hash,
{
    type MultiState = StateMachineBuilder<TState, TEvent>;
    type BuildState = StateMachineBuilder<TState, TEvent>;

    #[inline]
    fn transition_on(
        mut self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::MultiState, TState, TEvent> {
        self.add_transition_impl(self.current, event, next_state).map(|_| self)
    }

    #[inline]
    fn only_transition_on(
        mut self,
        event: TEvent,
        next_state: TState,
    ) -> Result<Self::BuildState, TState, TEvent>
    {
        self.add_transition_impl(self.current, event, next_state)
            .map(|_| self as StateMachineBuilder<TState, TEvent>)
    }
}
