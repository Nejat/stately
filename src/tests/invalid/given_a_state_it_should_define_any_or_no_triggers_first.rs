use stately::builder::{InitialState, Result, StateMachineBuilder};

fn main() -> Result<(), State, Event> {
    let _fsm = <StateMachineBuilder<State, Event>>::new(State::Initial)
        .add_start_state(Event::Start, State::Started).unwrap()
            .only_transition_on(Event::Next, State::Started).unwrap()
        .build()?;

    Ok(())
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum Event {
    Start,
    Next,
}

#[derive(Copy, Clone, Default, Debug, Eq, PartialEq, Hash)]
enum State {
    #[default]
    Initial,
    Started,
}
