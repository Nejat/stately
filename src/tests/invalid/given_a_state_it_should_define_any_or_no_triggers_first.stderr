error[E0599]: no method named `only_transition_on` found for associated type `<impl InitialState<State, Event> as InitialState<State, Event>>::TriggerState` in the current scope
 --> src/tests/invalid/given_a_state_it_should_define_any_or_no_triggers_first.rs:6:14
  |
6 |             .only_transition_on(Event::Next, State::Started).unwrap()
  |              ^^^^^^^^^^^^^^^^^^ method not found in `<impl InitialState<State, Event> as InitialState<State, Event>>::TriggerState`
